artemis-commons/src/test/java/org/apache/activemq/artemis/utils/ByteUtilTest.java

```diff
@@ -39,6 +39,49 @@
 
    private static Logger log = Logger.getLogger(ByteUtilTest.class);
 
+
+   @Test
+   public void testMixIDs1() {
+      long mixedID = ByteUtil.mixByteAndLong((byte)3, -1L);  // just a note, -1 is 0Xffffff....
+      Assert.assertEquals((byte)3, ByteUtil.getFirstByte(mixedID));
+
+      long mixedIDWithMaxValue = ByteUtil.mixByteAndLong((byte)3, Long.MAX_VALUE);
+      // first byte is 03, all the rest is fff...
+      Assert.assertEquals(0x03ffffffffffffffL, mixedIDWithMaxValue);
+      // -1 and fff should return you the same value when mixing 03, I'm doing this check just in case.
+      Assert.assertEquals(mixedIDWithMaxValue, mixedID);
```

Comment:
the other option would be to store two objects on the queue, and make a slow operation to acks. 

I don't think it's feasible here.

Reply:
@gemmellr to get two IDs mixing up on the same server you would need more than 0xffffffffffffffL messages duplicated on the same broker.

Reply:
You would only need 2 messages with IDs in that range to actually cause issues though. Yes, its a very big number, but just dividing it by 256 and silently assuming we can use those bits just seems fragile and likely to cause issues later. Hard to find and fix issues. At the very least I'd expect to see related changes elsewhere ensuring the broker can never actually generate such IDs, as presumably right now it can.

Looking, the ID actually seems to be down to the store to decide rather than the broker as a whole, so who is to say someone wont make a store that doesnt assign IDs differently, even just starting nearer the end to verify rollover is handled (something I've seen done elsewhere), or that its impossible a system will ever generate that ID for some strange less typical reason.

Actually, it even seems from the code like the current journal ID handling could potentially skip the number forward itself by up to 2^31-1 at a time with its checkpointing behaviour. If so, that will burn those IDs fairly quickly. Confirmed, I just burned several billion IDs in seconds while sending a mere handful of non-persistent messages, by simply killing the broker. Looks like it might do the same in other cases.

Reply:
this is moot now.